### PR TITLE
[fix] Auto serial no fecthed on the invoice even if stock update is disabled issue

### PR DESCRIPTION
**Pull request by rohitwaghchaure**

### CODE REVIEW

File: erpnext/controllers/accounts_controller.py [hunk 1]
```diff
@@ -192,11 +192,13 @@ def set_missing_item_details(self, for_validate=False):
...
 								item.set(fieldname, value)
 
-							elif fieldname == "cost_center" and not item.get("cost_center"):
+							elif fieldname in ['cost_center', 'conversion_factor'] and not item.get(fieldname):
 								item.set(fieldname, value)
 
-							elif fieldname == "conversion_factor" and not item.get("conversion_factor"):
-								item.set(fieldname, value)
+							elif fieldname == "serial_no":
+								stock_qty = item.get('stock_qty') * -1 if item.get('stock_qty') < 0 else item.get('stock_qty')
```
Comment (nabinhait) [inline erpnext/controllers/accounts_controller.py]: `stock_qty` field does not exists in Purchase Invoice

File: erpnext/public/js/controllers/transaction.js [hunk 2]
```diff
@@ -762,7 +763,8 @@ erpnext.TransactionController = erpnext.taxes_and_totals.extend({
 			"ignore_pricing_rule": me.frm.doc.ignore_pricing_rule,
 			"doctype": me.frm.doc.doctype,
 			"name": me.frm.doc.name,
-			"is_return": cint(me.frm.doc.is_return)
+			"is_return": cint(me.frm.doc.is_return),
+			"update_stock": (me.frm.doc.doctype == 'Sales Invoice') ? cint(me.frm.doc.update_stock) : 0
```
Comment (nabinhait) [inline erpnext/public/js/controllers/transaction.js]: also applicable to purchase invoice?

File: erpnext/stock/get_item_details.py [hunk 3]
```diff
@@ -172,10 +172,19 @@ def get_basic_details(args, item):
 		"net_amount": 0.0,
 		"discount_percentage": 0.0,
 		"supplier": item.default_supplier,
+		"update_stock": args.get("update_stock") if args.get('doctype') == 'Sales Invoice' else 0,
```
Comment (nabinhait) [inline erpnext/stock/get_item_details.py]: PI?

File: erpnext/stock/get_item_details.py [hunk 4]
```diff
@@ -532,6 +541,9 @@ def get_serial_no(args):
 		args = json.loads(args)
 		args = frappe._dict(args)
 
+	if args.get('doctype') == 'Sales Invoice' and not args.get('update_stock'):
```
Comment (nabinhait) [inline erpnext/stock/get_item_details.py]: PI?

File: erpnext/stock/get_item_details.py [hunk 5]
```diff
@@ -172,10 +172,19 @@ def get_basic_details(args, item):
 		"net_amount": 0.0,
 		"discount_percentage": 0.0,
 		"supplier": item.default_supplier,
+		"update_stock": args.get("update_stock") if args.get('doctype') == 'Sales Invoice' else 0,
 		"delivered_by_supplier": item.delivered_by_supplier if args.get("doctype") in ["Sales Order", "Sales Invoice"] else 0,
 		"is_fixed_asset": item.is_fixed_asset
 	})
 
+	if out.uom != args.uom:
```
Comment (nabinhait) [inline erpnext/stock/get_item_details.py]: always set conversion_factor and stock_qty, even if sales uom and stock uom is same.

**Review by nabinhait — CHANGES_REQUESTED** [review]

File: erpnext/controllers/accounts_controller.py [hunk 6]
```diff
@@ -192,11 +192,14 @@ def set_missing_item_details(self, for_validate=False):
...
 								item.set(fieldname, value)
 
-							elif fieldname == "cost_center" and not item.get("cost_center"):
+							elif fieldname in ['cost_center', 'conversion_factor'] and not item.get(fieldname):
 								item.set(fieldname, value)
 
-							elif fieldname == "conversion_factor" and not item.get("conversion_factor"):
-								item.set(fieldname, value)
+							elif fieldname == "serial_no":
+								qty_field = "qty" if ret.doctype == 'Purchase Invoice' else "stock_qty"
```
Comment (nabinhait) [inline erpnext/controllers/accounts_controller.py]: Can you fix this, as now we have stock_qty field in Purchase Invoice Item?

**Review by nabinhait — CHANGES_REQUESTED** [review]

File: erpnext/controllers/accounts_controller.py (hunk 6 above)
Comment (rohitwaghchaure) [inline erpnext/controllers/accounts_controller.py]: Fixed

**Review by nabinhait — APPROVED** [review]

File: erpnext/patches.txt [hunk 7]
```diff
@@ -384,4 +384,5 @@ erpnext.patches.v8_0.repost_reserved_qty_for_multiple_sales_uom
 erpnext.patches.v8_0.addresses_linked_to_lead
 execute:frappe.delete_doc('DocType', 'Purchase Common')
 erpnext.patches.v8_0.update_stock_qty_value_in_purchase_invoice
-erpnext.patches.v8_0.update_supplier_address_in_stock_entry
\ No newline at end of file
+erpnext.patches.v8_0.update_supplier_address_in_stock_entry
+execute:frappe.db.sql("update `tabSales Invoice Item` set serial_no = '' where parent in (select name from `tabSales Invoice` where update_stock = 0 and docstatus = 1)")
```
Comment (nabinhait) [inline erpnext/patches.txt]: Put this patch in a separate file. Anything lengthy should be inside a file.

**Review by nabinhait — CHANGES_REQUESTED** [review]

File: erpnext/patches.txt (hunk 7 above)
Comment (rohitwaghchaure) [inline erpnext/patches.txt]: Fixed!

**Review by nabinhait — APPROVED** [review]

**Pull request merged**